src/challenge/ChallengeContext.ts

```diff
@@ -460,6 +479,43 @@ class ChallengeContextClass {
           }
         });
     },
+    "fetch-file": (
+      filename: string,
+      store: EditableBookStore | null = null
+    ) => {
+      this.challenge.props.fetcher
+        .fetch(filename, this.challenge.props.authContext)
+        .then((response) => {
+          if (!response.ok) {
+            // try to make the file
+            let newContents = "add new file contents here\n";
+            if (store) {
+              store.store.save(newContents, `edit://edit/${filename}`);
+            } else {
+              throw Error("file not available");
+            }
+            return newContents;
+          }
+          return response.text();
+        })
+        .then((text) => {
+          this.challenge.additionalFilesLoadCallback(filename, text);
```

Comment:
yeah, it would be definitely neater to return a promise, so this additionalFilesLoadCallback can be gone. Promises naturally allow you to handle the success and the fail paths